porch/controllers/rootsyncsets/pkg/controllers/rootsyncset/controller.go

```diff
@@ -108,59 +148,246 @@ func (r *RootSyncSetReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+	var sb strings.Builder
+	for clusterRef, res := range r {
+		if res.clientError != nil {
+			sb.WriteString(fmt.Sprintf("failed to create client for %s: %v\n", clusterRef, res.clientError))
+		}
+		if res.patchError != nil {
+			sb.WriteString(fmt.Sprintf("failed to patch %s: %v\n", clusterRef, res.patchError))
+		}
+		if res.statusError != nil {
+			sb.WriteString(fmt.Sprintf("failed to check status for %s: %v\n", clusterRef, res.statusError))
+		}
+	}
+	return sb.String()
+}
+
+type clusterRefResult struct {
+	clientError error
+	patchError  error
+	statusError error
+	status      string
+}
+
+// patchRootSync patches the RootSync in the remote clusters targeted by
+// the clusterRefs based on the latest revision of the template in the RootSyncSet.
+func (r *RootSyncSetReconciler) patchRootSync(ctx context.Context, client dynamic.Interface, name string, rss *v1alpha1.RootSyncSet) error {
+	newRootSync, err := BuildObjectsToApply(rss)
 	if err != nil {
-		return schema.GroupVersionResource{}, nil, fmt.Errorf("failed to parse group version when building object: %w", err)
+		return err
+	}
+	data, err := json.Marshal(newRootSync)
+	if err != nil {
+		return fmt.Errorf("failed to encode root sync to JSON: %w", err)
+	}
+	rs, err := client.Resource(rootSyncGVR).Namespace(rootSyncNamespace).Patch(ctx, name, types.ApplyPatchType, data, metav1.PatchOptions{FieldManager: name})
+	if err != nil {
+		return fmt.Errorf("failed to patch RootSync: %w", err)
 	}
-	rootSyncRes := schema.GroupVersionResource{Group: gv.Group, Version: gv.Version, Resource: "rootsyncs"}
+	klog.Infof("Create/Update resource %s as %v", name, rs)
+	return nil
+}
+
+// setupWatches makes sure we have the necessary watches running against
+// the remote clusters we care about.
+func (r *RootSyncSetReconciler) setupWatches(ctx context.Context, client dynamic.Interface, rssName, ns string, clusterRef v1alpha1.ClusterRef) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	nn := types.NamespacedName{
+		Namespace: ns,
+		Name:      rssName,
+	}
+
+	// If we already have a watch running, make sure we have the current RootSyncSet
+	// listed in the liens map.
+	if w, found := r.watchers[clusterRef]; found {
+		w.liens[nn] = struct{}{}
+		return
+	}
+
+	// Since we don't currently have a watch running, create a new watcher
+	// and add it to the map of watchers.
+	watcherCtx, cancelFunc := context.WithCancel(context.Background())
+	w := &watcher{
+		clusterRef: clusterRef,
+		ctx:        watcherCtx,
+		cancelFunc: cancelFunc,
+		client:     client,
+		channel:    r.channel,
+		liens:      make(map[types.NamespacedName]struct{}),
+	}
+	go w.watch()
+	r.watchers[clusterRef] = w
+}
+
+// pruneWatches removes the current RootSyncSet from the liens map of all watchers
+// that it no longer needs. If any of the watchers are no longer used by any RootSyncSets,
+// they are shut down.
+func (r *RootSyncSetReconciler) pruneWatches(rssnn types.NamespacedName, clusterRefs []*v1alpha1.ClusterRef) {
```

Comment:
I do wonder if this could be combined with `setupWatches`, but we can evolve

Reply:
Yeah, it might. I added it separately as pruning watches requires looking at all clusterRefs together, while setting them up fits well within the for loop for getting the client and doing the patching.